src/GoTrueClient.ts

```diff
@@ -380,25 +380,112 @@ export default class GoTrueClient {
     }
   }
 
+  /**
+   * Sets the session from the provided session information. The access_token
+   * is reused if it is not expired, otherwise a new access token is fetched by
+   * refreshing the session with the provided refresh_token.
+   *
+   * This method is useful when using in a server-side rendered context.
+   *
+   * @param params.refresh_token A valid refresh token (typically obtained from a cookie)
+   * @param params.access_token An access token (typically obtained from a cookie)
+   * @param params.provider_token A provider token (from social login, for example), optional
+   * @param params.expires_in A timeout in seconds when the access token will expire, if not provided taken from the `exp` claim in the access token
+   */
+  async setSession(params: {
+    refresh_token: string
+    access_token: string
+    expires_in?: number
+    provider_token?: string
+  }): Promise<{ session: Session | null; error: ApiError | null }>
+
   /**
    * Sets the session data from refresh_token and returns current Session and Error
    * @param refresh_token a JWT token
    */
   async setSession(
     refresh_token: string
-  ): Promise<{ session: Session | null; error: ApiError | null }> {
-    try {
-      if (!refresh_token) {
-        throw new Error('No current session.')
-      }
+  ): Promise<{ session: Session | null; error: ApiError | null }>
+
+  async setSession(arg0: any): Promise<{ session: Session | null; error: ApiError | null }> {
+    let session: Session
+
+    if (typeof arg0 === 'string') {
+      // using the refresh_token string API
+      const refresh_token = arg0
+
       const { data, error } = await this.api.refreshAccessToken(refresh_token)
       if (error) {
         return { session: null, error: error }
       }
 
-      this._saveSession(data!)
+      session = data!
+    } else if (
+      typeof arg0 === 'object' &&
+      arg0 &&
+      typeof arg0.refresh_token === 'string' &&
+      typeof arg0.access_token === 'string' &&
+      (typeof arg0.expires_in === 'number' || !arg0.expires_in) &&
+      (typeof arg0.provider_token === 'string' || !arg0.provider_token)
+    ) {
+      // using the object parameter API
+      let { provider_token, refresh_token, access_token, expires_in } = arg0
+
+      const timeNow = Math.round(Date.now() / 1000)
+      let expires_at = 0
+
+      if (expires_in) {
+        expires_at = timeNow + expires_in
```

Comment:
why should we allow passing in an `expires_in` time - seems like this is never validated against the `exp` claim in the jwt

Reply:
It is validated later on line 406:

```typescript
      if (timeNow > expires_at) {
```

Reply:
isn't `timeNow` just the current time and not the `exp` claim in the jwt? if `expires_in` is passed, then `timeNow` will remain as the current time since line 453 never happens